radfield.cc

```diff
@@ -1149,9 +1149,15 @@ void reduce_estimators()
 
   if constexpr (DETAILED_BF_ESTIMATORS_ON) {
     for (ptrdiff_t nonemptymgi = 0; nonemptymgi < nonempty_npts_model; nonemptymgi++) {
-      MPI_Allreduce(MPI_IN_PLACE, &bfrate_raw[nonemptymgi * globals::bfestimcount], globals::bfestimcount, MPI_DOUBLE,
-                    MPI_SUM, MPI_COMM_WORLD);
+      MPI_Reduce(globals::rank_in_node == 0 ? MPI_IN_PLACE : &bfrate_raw[nonemptymgi * globals::bfestimcount],
+                 &bfrate_raw[nonemptymgi * globals::bfestimcount], globals::bfestimcount, MPI_DOUBLE, MPI_SUM, 0,
+                 globals::mpi_comm_node);
     }
+    if (globals::rank_in_node == 0) {
+      MPI_Allreduce(MPI_IN_PLACE, bfrate_raw.data(), nonempty_npts_model * globals::bfestimcount, MPI_DOUBLE, MPI_SUM,
+                    globals::mpi_comm_internode);
+    }
+    MPI_Bcast(bfrate_raw.data(), nonempty_npts_model * globals::bfestimcount, MPI_DOUBLE, 0, globals::mpi_comm_node);
```

Comment:
Sorry, this just occured to me. An MPI_Allreduce followed by an MPI_Bcast should be less efficient than an MPI_Allreduce. Would it be better to replace the loop of MPI_Allreduce calls over each cell with just one since the entire array needs to be reduced? This code is there because it used to selectively reduce only non-empty cells and skip over empty ones, but that's not needed now that only non-empty cells get stored in the array.

Reply:
Actually, there might be a reason to keep that loop. Sending a size of (nonempty_npts_model * globals::bfestimcount) could overflow the 32-bit int size parameter of the MPI call.